Update bundle to support Flysystem v2

Lustmored opened the pull request:
I have built upon flysystem-v2 branch and set it as a target as probably such a change will require major version bump.

Most of the work was updating tests to actually run against new architecture. I have commented and removed functionality and adapters that are missing from v2.

I haven't touched docs and didn't add support for any new options as I wanted to get feedback before eventually delving more.

This PR also includes merging master into flysystem-v2 as a base for work.

tgalopin commented:
Thanks a lot @Lustmored !

It seems you should probably rebase on master as the first commits in your PR are newly merged commits.

I'll review this asap :) .

Lustmored commented:
> Thanks a lot @Lustmored !
> 
> It seems you should probably rebase on master as the first commits in your PR are newly merged commits.
> 
> I'll review this asap :) .

Right! I have force-pushed rebase onto master.

tgalopin commented:
There still seems to be a merge problem, are you sure you rebased properly?

Lustmored commented:
I am pretty sure. Yet as I've stated - this PR is based on 'flysystem-v2' branch from original repository and targets it. I believe rebase of 'flysystem-v2' branch onto master will fix the conflicts. Alternatively I can target 'master' branch.

tgalopin commented:
Ah right! My bad, I must have badly rebased myself then :) I'll have a look asap (I'm preparing the SymfonyWorld conference right now, kinda busy but I'll have more time later :) )

Lustmored commented:
I have changed PR target to `master` for easier review and merge.

I have also updated docs and re-added cached storage driver based on `lustmored/flysystem-v2-simple-cache-adapter` for easier upgrade path with useful, yet lighter, cached filesystem driver. If you'd prefer I don't just let me know and I will drop that part from the PR :)

tgalopin commented:
Thanks for the ping, I'll have a look :)

Lustmored commented:
@tgalopin Just a friendly remainder, no pressure intended :)

I've found out my commits authoring was off, so fixed it and force pushed. No changes in code this time.

tgalopin commented on composer.json:
```diff
@@ -20,36 +20,24 @@
     "require": {
-        "php": ">=7.1",
-        "league/flysystem": "^1.0.40",
+        "php": "^7.2 || ^8.0",
```
Please keep the `>=` notation, it's easier to maintain.

tgalopin commented on composer.json:
```diff
@@ -20,36 +20,24 @@
-        "league/flysystem-ziparchive": "^1.0",
+        "league/flysystem-aws-s3-v3": "^2.0",
+        "league/flysystem-memory": "^2.0",
+        "lustmored/flysystem-v2-simple-cache-adapter": "^0.1.0",
```
There is no cache adapter anymore? Perhaps would it make sense to remove the adapter then?

Lustmored replied:
There is no upstream adapter for v2 and there seems to be no work towards it. V1 adapter also was very memory inefficient (https://github.com/thephpleague/flysystem-cached-adapter/issues/23). That is why I have created this simple alternative for v2, which is very helpful in my projects already, so I think it can be helpful for others too.

But of course if you feel like dropping it from the bundle, I can do so.

tgalopin replied:
Let's drop it for now, we'll add it if someone find it useful (not sure it is)

Lustmored replied:
OK, I have removed it from code and docs

tgalopin commented on docs/2-cloud-storage-providers.md:
```diff
@@ -4,37 +4,10 @@ One of the core feature of Flysystem is its ability to interact easily with remo
 * [AWS S3](#aws-s3)
 * [DigitalOcean Spaces](#digitalocean-spaces)
 * [Scaleway Object Storage](#scaleway-object-storage)
-* [Google Cloud Storage](#google-cloud-storage)
```
We are removing quite useful adapters here (Azure and Google Cloud Storage), are there community alternatives?

Lustmored replied:
All Azure adapters doesn't even have issues for flysystem v2.

I just found that league published google cloud adapter recently (https://github.com/thephpleague/flysystem-google-cloud-storage) and I will work towards this adapter soon and will update PR.

tgalopin replied:
Let's implement the official Google Cloud Storage package yes, and let's ignore Azure for now: if people need it, it'll be added back.

Lustmored replied:
OK, I have already done that :)

tgalopin commented on docs/4-caching-metadata-in-symfony-cache.md:
```diff
@@ -10,19 +10,23 @@ involves an additional overhead on top of the classical slowness of I/O.
 improve performances on this level. To do this, you can use the `cache` adapter.
 
-> *Note:* this adapter caches anything but the file content. This keeps the cache 
+> *Note:* this adapter caches just the metadata of existing files. This keeps the cache 
```
this adapter only caches the metadata of existing files, not their content.

tgalopin commented on docs/4-caching-metadata-in-symfony-cache.md:
```diff
@@ -10,19 +10,23 @@ involves an additional overhead on top of the classical slowness of I/O.
+> *Note:* this adapter caches just the metadata of existing files. This keeps the cache 
 > small enough to be beneficial and covers all the file system inspection operations.
 
+> *Note:* Earlier versions used different cache driver that was memory heave on systems
```
Unnecessary note IMO

Lustmored replied:
OK, removed

tgalopin commented on docs/4-caching-metadata-in-symfony-cache.md:
```diff
@@ -10,19 +10,23 @@ involves an additional overhead on top of the classical slowness of I/O.
 
 The cache adapter works using a source storage (from which it will read the uncached data)
-and a [Symfony cache pool](https://symfony.com/doc/current/reference/configuration/framework.html#pools). 
+and any [PSR-6 cache item pool interface](https://www.php-fig.org/psr/psr-6/). 
```
A Symfony pool is a PSR-6 pool. We're making a Symfony bundle, let's keep the Symfony reference.

tgalopin commented on tests/Adapter/Builder/AsyncAwsAdapterDefinitionBuilderTest.php:
```diff
@@ -39,39 +39,39 @@ public function provideValidOptions()
-                'ServerSideEncryption' => 'AES256',
-            ],
-        ]];
+//        yield 'options' => [[
```
Either we should remove this code or enable it back

tgalopin commented on tests/Adapter/Builder/AsyncAwsAdapterDefinitionBuilderTest.php:
```diff
@@ -39,39 +39,39 @@ public function provideValidOptions()
-        $this->assertSame('prefix/path', $definition->getArgument(2));
-        $this->assertSame(['ServerSideEncryption' => 'AES256'], $definition->getArgument(3));
-    }
+//    public function testOptionsBehavior()
```
Same

tgalopin commented on tests/Adapter/options.yaml:
```diff
@@ -13,24 +13,24 @@ fs_aws:
-        client: 'azure_client_service'
-        container: 'container_name'
-        prefix: 'optional/path/prefix'
+#fs_azure:
```
Same and for the next commented code too

tgalopin commented on tests/Adapter/Builder/LocalAdapterDefinitionBuilderTest.php:
```diff
@@ -78,18 +78,18 @@ public function provideValidOptions()
             'file' => [
-                'public' => 0755,
-                'private' => 0755,
+                'public' => 0666,
```
This feels like a bug as the config didn't change

Lustmored replied:
Right, I was checking something during development and missed reverting this change.

tgalopin reviewed (commented):
Hi!

Thanks for the friendly ping :) . Here is a review. Also, you should rebase your branch on master to get the latest CI fixes.

More generally: I would find it useful to keep Azure and Google Cloud support. Are there credible community alternatives for Flysytem 2?

Lustmored commented:
I have updated PR with Google Cloud support, rebase to master and changes requested in comments

tgalopin commented:
Thanks so much @Lustmored ! I'll have a look later today but thanks for your work :) !

nCrazed commented:
@Lustmored, @tgalopin is there anything that you're missing, could use some help with here?

I was planning to add flysystem to an existing Symfony project, but decided to postpone until this bundle starts can support V2. I am happy to put in some work in order to speed it up

Lustmored commented:
> @Lustmored, @tgalopin is there anything that you're missing, could use some help with here?
> 
> I was planning to add flysystem to an existing Symfony project, but decided to postpone until this bundle starts can support V2. I am happy to put in some work in order to speed it up

We have to wait until @tgalopin has time to take another look and possibly merge and release

tgalopin commented:
Thanks so much for your work @Lustmored ! I'll release 2.0 soon.

tgalopin commented on tests/DependencyInjection/FlysystemExtensionTest.php:
```diff
@@ -31,19 +27,10 @@ public function provideFilesystems()
-            'fs_dropbox',
             'fs_ftp',
             'fs_gcloud',
-            'fs_lazy',
```
I missed this, why did you remove this?

tgalopin commented on tests/DependencyInjection/FlysystemExtensionTest.php:
```diff
@@ -31,19 +27,10 @@ public function provideFilesystems()
             'fs_aws',
-            'fs_azure',
-            'fs_cache',
-            'fs_custom',
```
I missed this, why did you remove this?

tgalopin commented on tests/Adapter/Builder/GcloudAdapterDefinitionBuilderTest.php:
```diff
@@ -31,23 +31,10 @@ public function provideValidOptions()
-            'prefix' => 'prefix/path',
-        ]];
-
-        yield 'api_url' => [[
```
I missed this, why did you remove this?

tgalopin commented on tests/Adapter/Builder/GcloudAdapterDefinitionBuilderTest.php:
```diff
@@ -31,23 +31,10 @@ public function provideValidOptions()
             'bucket' => 'bucket',
         ]];
 
-        yield 'prefix' => [[
```
I missed this, why did you remove this?

tgalopin commented on tests/Kernel/config/flysystem.yaml:
```diff
@@ -7,35 +7,13 @@ flysystem:
-                store: 'cache.app'
-                source: 'fs_local'
-
-        fs_custom:
```
I think we should keep this shouldn't we?

tgalopin reviewed (commented):
A bit late, I agree, but I just realized a few things

tgalopin commented:
Released

@nCrazed could you test to ensure everything is good on your side :) ? I tested on my projects but more tests can't hurt.

nCrazed commented:
I am planning to start the refactoring/integration next week. Will report any problems that I run into.

Lustmored commented:
@tgalopin @nCrazed if any issues arise I am more than willing to help tackling them. Please mention me in any issue related to this and I will gladly help.

tgalopin commented:
Thanks @Lustmored, really appreciate the help!

nspyke commented:
Hi @tgalopin 
I'm currently using this bundle and the AWS adapter. I need to move my app to Azure. Would it be possible to reinstate Azure Blob storage support?
I'd be happy to submit a PR if it were to be reviewed and merged,
